### PR TITLE
kvs: add flux kvs ls subcommand

### PR DESCRIPTION
**Pull request by garlick**
This adds a `flux kvs ls` subcommand that behaves similar to ls(1), as discussed in #1158.

```
Usage: flux-kvs ls [-R] [-d] [-w COLS] [-1] [-W] [-C] [key...]
List directory
  -1, --1                Force one entry per line
  -C, --columns          List entries by columns
  -F, --classify         Append indicator (one of .@) to entries
  -R, --recursive        List directory recursively
  -d, --directory        List directory instead of contents
  -h, --help             Display this message.
  -w, --width            Set output width to COLS.  0 means no limit
```

### CODE REVIEW

**Comment by garlick** [comment]
Restarted three tests; two write errors, one hydra test failure (can't really be related).

**Comment by garlick** [comment]
Rebased after merge of #1171

File: src/cmd/flux-kvs.c [hunk 1]
```diff
@@ -816,6 +851,254 @@ int cmd_dir (optparse_t *p, int argc, char **argv)
...
+    return max;
+}
+
+/* Determine appropriate terminal window width for columnar output.
+ * The options -w COLS or -1 force the width to COLS or 1, respectively.
+ * If not forced, ask the tty how wide it is.  If not on a tty, assume 80
+ * if -C, or 1 if not.
+ * N.B. A width of 0 means "unlimited", while a width of 1 effectively
+ * forces one entry per line since entries are never broken across lines.
+ */
```
Comment (grondo) [inline src/cmd/flux-kvs.c]: To match `ls` should `get_window_width()` also check `COLUMNS` env var? From GNU `ls(1)`:
``` 
‘-w cols’
‘--width=cols’
Assume the screen is cols columns wide. The default is taken from the terminal settings if possible; otherwise the environment variable COLUMNS is used if it is set; otherwise the default is 80. With a cols value of ‘0’, there is no limit on the length of the output line, and that single output line will be delimited with spaces, not tabs.
```

**Comment by grondo** [comment]
I know the `-C` option is coming from `ls`, but is there a use case for it in `flux-kvs ls`? From the cmdline help and manpage it sounds like a superfluous option. Is it meant to be used when output is not a terminal? If it has some use case a note in the manual might help.

File: src/cmd/flux-kvs.c [hunk 2]
```diff
@@ -116,6 +143,13 @@ static struct optparse_subcommand subcommands[] = {
       0,
       dir_opts
     },
+    { "ls",
+      "[-R] [-d] [-w COLS] [-1] [-W] [-C] [key...]",
+      "List directory",
```
Comment (chu11) [inline src/cmd/flux-kvs.c]: the ```-W``` iss probably supposed to be ```-F```?

Comment (chu11) [inline src/cmd/flux-kvs.c]: also noticed the manpage is [-R] [-d] [-C] [-1|-C] [-w COLS] [-F], should this match up together (most notably [-1|-C].

File: src/cmd/flux-kvs.c [hunk 3]
```diff
@@ -816,6 +850,254 @@ int cmd_dir (optparse_t *p, int argc, char **argv)
...
+        }
+        kvsdir_destroy (dir);
+    }
+done:
+    flux_future_destroy (f);
+}
+
+/* List keys, arranging in columns so that they fit within 'win_width',
+ * and using a custom column width selected based on the longest entry name.
+ * The keys are popped off the list and freed we go.
```
Comment (chu11) [inline src/cmd/flux-kvs.c]: ```freed we go```, just "freed"?  Or had another thought but didn't finish?

File: src/cmd/flux-kvs.c (hunk 1 above)
Comment (garlick) [reply src/cmd/flux-kvs.c]: Good catch, I did not read that carefully in the ls(1) page.  Will fix.

File: src/cmd/flux-kvs.c (hunk 2 above)
Comment (garlick) [reply src/cmd/flux-kvs.c]: Yep, thanks.

Comment (garlick) [reply src/cmd/flux-kvs.c]: Indeed.

File: src/cmd/flux-kvs.c (hunk 3 above)
Comment (garlick) [reply src/cmd/flux-kvs.c]: Freed as we go, but I'll go with "freed".

**Comment by garlick** [comment]
> I know the -C option is coming from ls, but is there a use case for it in flux-kvs ls? From the cmdline help and manpage it sounds like a superfluous option. Is it meant to be used when output is not a terminal? If it has some use case a note in the manual might help.

I may have misrepresented the way columns are selected by ls(1), and thought -C was needed to get columns when output is not a tty.  I think we could just drop it actually.

**Comment by grondo** [comment]
> I may have misrepresented the way columns are selected by ls(1), and thought -C was needed to get columns when output is not a tty. I think we could just drop it actually.

Ah, ok that is what I figured. I see you already removed it, but I would have been fine with keeping `-C`, however a note in the usage message "`preserve column output when stdout is not a tty`" I think would have avoided confusion. In the real `ls` there are so many options that I think the `-C` just gets lost in the noise, so isn't confusing...

**Comment by grondo** [comment]
One missing `kvsitr_destroy` in `flux-kvs.c:list_kvs_dir()`:
(Actually, could any iterators created on a kvsdir be automatically destroyed when the dir is destroyed? Not sure if this is bad practice though...)
```diff
--- a/src/cmd/flux-kvs.c
+++ b/src/cmd/flux-kvs.c
@@ -976,6 +976,7 @@ static void list_kvs_dir (flux_t *h, const char *key, optpar
                 free (nkey);
             }
         }
+        kvsitr_destroy (itr);
         kvsdir_destroy (dir);
     }
 done:
```

**Comment by grondo** [comment]
I did poke at this and it seemed to work great!

One thing I did notice is that `flux kvs ls resource` works, but `flux kvs ls resource.` returns `resource.: No such file or directory`. However, the same thing happens with `flux-kvs dir` so I'm not sure it is problem

**Comment by garlick** [comment]
Hmm, yeah, ls(1) allows `ls foo/` so probably flux-ls should allow `flux ls foo.`

Good catch on the `kvsitr_destroy()`.  The kvsdir also still needs to be destroyed but not in the conditional.  A few patches coming...

**Comment by grondo** [comment]
> Hmm, yeah, ls(1) allows ls foo/ so probably flux-ls should allow flux ls foo.
One catch will be that `flux kvs ls foo.` should fail with `Not a directory` if `foo` is not a directory.

Nice work, `flux kvs ls` feels much more natural than `flux kvs dir`

**Comment by chu11** [comment]
> Hmm, yeah, ls(1) allows ls foo/ so probably flux-ls should allow flux ls foo.

I noticed this a long time ago and just accepted it as the way flux kvs worked, but it does seem like it should be supported.  The behavior should also be in ```flux kvs get```.

```
>flux kvs get resource.
flux-kvs: resource.: No such file or directory
>flux kvs get resource
flux-kvs: resource: Is a directory
```

Presumably, the error should be identical.

This could be a different issue and maybe we should just open a issue on it.  In fact, the solution may have to come inside the flux kvs module as that is what is returning some of these error codes.

**Comment by grondo** [comment]
> This could be a different issue and maybe we should just open a issue on it. In fact, the solution may have to come inside the flux kvs module as that is what is returning some of these error codes.

I'm fine with fixing this holistically in another PR

**Comment by garlick** [comment]
Good point @chu11 - want to open that issue?

**Comment by chu11** [comment]
One thing I noticed when messing with this branch

```
>flux kvs ls -FR .
<snip>
resource.hwloc:
by_rank.   xml.       loaded.    

resource.hwloc.by_rank:
2.   0.   1.   3.   
```

Do we want entries sorted before it is output?  Because ```ls``` does as well.  Dunno if it would come within the ```kvsdir``` api or not.  It appears that ```jansson``` does not sort on iterators, so it'd have to be some wrapper intermediary layer as well.  This could be for a future issue and not necessarily in this PR.

**Comment by garlick** [comment]
> Do we want entries sorted before it is output?

We could just use the `JSON_SORT_KEYS` flag when creating a kvsdir_t.  And the list of loose keys can also be sorted with `zlist_sort()`.

Sounds like a good idea.

File: src/cmd/flux-kvs.c [hunk 4]
```diff
@@ -816,6 +847,304 @@ int cmd_dir (optparse_t *p, int argc, char **argv)
...
+{
+    flux_t *h = (flux_t *)optparse_get_data (p, "flux_handle");
+    flux_future_t *f;
+    const char *json_str;
+    char *nkey;
+    json_t *treeobj = NULL;
+    bool require_directory = false;
+
+    if (!(nkey = malloc (strlen (key) + 2))) // room for deocration char + null
+        log_err_exit ("malloc");
```
Comment (chu11) [inline src/cmd/flux-kvs.c]: typo, "decoration" :-)

Comment (garlick) [reply src/cmd/flux-kvs.c]: oops :-/

**Comment by garlick** [comment]
Output is sorted now.   Shall I squash this down or is it still useful for review purposes to see the incremental work?

**Comment by chu11** [comment]
I think it's ok to squash now.  One nit, the ```| sort``` can probably be removed in the tests now.

**Comment by garlick** [comment]
Heading out for a bit so having addressed the recent travis failures, I'm going to go ahead and squash.

File: t/t1000-kvs.t [hunk 5]
```diff
@@ -279,6 +279,123 @@ EOF
...
+	flux kvs ls -d $DIR.a. >output &&
+	cat >expected <<-EOF &&
+	$DIR.a
+	EOF
+	test_cmp expected output
+'
+test_expect_success 'kvs: ls key. fails if key is not a directory' '
+	flux kvs unlink -Rf $DIR &&
+	flux kvs put $DIR.a=42 &&
+	! flux kvs ls -d $DIR.a.
```
Comment (grondo) [inline t/t1000-kvs.t]: Just a general comment, no reason to fix it here and now

Instead of `! command` to indicate expected (required) failure, it might be more correct to use either `test_must_fail` or `test_expect_code`. See sharness docs [here](https://github.com/chriscool/sharness/blob/master/API.md#test_must_fail)

**Comment by garlick** [comment]
I'll go ahead and address the unnecessary sort and the use of `!` in the test.

**Comment by grondo** [comment]
Thanks, LGTM.

File: src/common/libkvs/kvs_dir.c [hunk 6]
```diff
@@ -140,18 +140,45 @@ const char *kvsdir_rootref (kvsdir_t *dir)
...
+}
+
 kvsitr_t *kvsitr_create (kvsdir_t *dir)
 {
-    kvsitr_t *itr;
+    kvsitr_t *itr = NULL;
+    const char *key;
+    json_t *dirdata, *value;
 
+    if (!dir) {
```
Comment (chu11) [inline src/common/libkvs/kvs_dir.c]: Should we add a unit test for ```kvsitr_create()``` now that it's checking for NULL input?

File: src/common/libkvs/kvs_dir.c [hunk 7]
```diff
@@ -160,16 +187,21 @@ kvsitr_t *kvsitr_create (kvsdir_t *dir)
...
 }
 
 const char *kvsitr_next (kvsitr_t *itr)
 {
     const char *name = NULL;
 
-    if (itr->iter) {
-        name = json_object_iter_key (itr->iter);
-        itr->iter = json_object_iter_next (itr->dirdata, itr->iter);
+    if (itr) {
```
Comment (chu11) [inline src/common/libkvs/kvs_dir.c]: possibly a unit test for bad input here as well? although it might not be necessary coverage wise for this one.

**Comment by garlick** [comment]
Sure, I'll do that.

**Comment by garlick** [comment]
I added the tests you suggested to the commit that modified the `kvsitr_t` implementation (squashed in and forced).

**Comment by chu11** [comment]
LGTM, if no complaints will hit button.
